OpenID connect chapter and chapter template

prestwick opened the pull request:
- [X ] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/systemlink-operations-handbook/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?
Adds a chapter for setting up OpenID connect configuration in SystemLink. It also creates a chapter-template that can be referenced for future documentation. 

### Why should this Pull Request be merged?

Provide supporting docs in OIDC feature in SL 2020R4

### What testing has been done?
Ensure everything renders correctly in MKdocs. Check links.

chris468 commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,133 @@
+
+> For example, a OpenID provider with the DNS `openid-provider.com` would yield files named `openid-provider.com/conf` , `openid-provider.com.client`, and `openid-provider.com.provider`. 
+
+These files do no exist for new SystemLink installations. Each file must be added to the `C:\Program Files\National Instruments\Shared\Web Server\conf\openidc` directory. The NI Web Server must be restarted for these changes to take effect.
```
```suggestion
These files do not exist for new SystemLink installations. Each file must be added to the `C:\Program Files\National Instruments\Shared\Web Server\conf\openidc` directory. The NI Web Server must be restarted for these changes to take effect.
```

prestwick replied:
fixed!

cameronwaterman commented on handbook/openid-connect/openid-connect.md:
````diff
@@ -0,0 +1,133 @@
+```
+
+### Setting Login Redirect URI
+The client configuration for your provider requires a redirect URL that is during the login flow. This must include the protocol (`https://` or `http:// `) and the DNS name of the SystemLink server. If the SystemLink server's DNS changes, this setting must be updated with the provider. 
````
"that is during the login flow" sounds kind of weird to me

prestwick replied:
missing word "used". Fixed!

chris468 commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,133 @@
+
+These files do no exist for new SystemLink installations. Each file must be added to the `C:\Program Files\National Instruments\Shared\Web Server\conf\openidc` directory. The NI Web Server must be restarted for these changes to take effect.
+
+SystemLink can be configured to support multiple OpenID Providers simultaneously by creating a `[provider-dns].conf`, `[provider-dns].client`, and `[provider-dns].provider` file for each provider.  It is required that the user ID used within SystemLink be unique across providers. This id take the form `[sub_claim]@issuer` where this issuer is the DNS name of the OpenID Connect provider. You can s ...
```
```suggestion
SystemLink can be configured to support multiple OpenID Providers simultaneously by creating a `[provider-dns].conf`, `[provider-dns].client`, and `[provider-dns].provider` file for each provider.  It is required that the user ID used within SystemLink be unique across providers. This id takes the form `[sub_claim]@issuer` where this issuer is the DNS name of the OpenID Connect provider. You can see the id SystemLink uses for a user by viewing the details of that user in the SystemLink security application. 
```

prestwick replied:
fixed!

cameronwaterman commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,133 @@
+If a claim returns an array of values any one of these values can be supplied in the **Value** field to create a mapping. 
+
+## Troubleshooting Failed Authentication
+Due to the amount of configuration required by both the provider and SystemLink its can be challenging to enable the entire OpenID Connect flow. The following sources can be used to troubleshoot a failed connection. 
```
"its" - remove 's'

prestwick replied:
fixed!

cameronwaterman commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,133 @@
+
+**OpenID Connect Provider logs:** Consult your OpenID Connect Provider's documentation on the location of their application log files. 
+
+**NI Web Server Logs:** These are found at `C:\ProgramData\National Instruments\Web Server\logs\error.log`
```
Related to the email Chris sent out yesterday, it might be worth clarifying that the suffixed version of this file oftentimes contains the latest logs. Users familiar with how Linux log files work might find this confusing.

prestwick replied:
I checked into this and the log without the prefix was always the latest. This feels like a bug if that is not the case. Especially if it doesn't seem like the behavior is always the same.

chris468 replied:
I saw it using files with a suffix yesterday when I sent the email.  @spanglerco had mentioned that it's the way the apache rotate log tool works on windows.

prestwick replied:
My recommendation is to file a bug and resolve this issue rather than try and document around it here.

cameronwaterman reviewed (approved):
Mostly nit-picky comments regarding grammar. Some of my suggestions might not be correct so feel free to ignore anything you think is incorrect.

chris468 commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,133 @@
+
+> For example, a OpenID provider with the DNS `openid-provider.com` would yield files named `openid-provider.com/conf` , `openid-provider.com.client`, and `openid-provider.com.provider`. 
+
+These files do no exist for new SystemLink installations. Each file must be added to the `C:\Program Files\National Instruments\Shared\Web Server\conf\openidc` directory. The NI Web Server must be restarted for these changes to take effect.
```
I believe the web server must be restarted for changes to any of the 3 files to take effect, not just this one.

prestwick replied:
I clarified this.

chris468 commented on handbook/openid-connect/openid-connect.md:
````diff
@@ -0,0 +1,133 @@
+ }
+}
+```
+In this example the `openid`, `email`, and `profile` scopes are requested. Additional scopes may be requested. Consult your provider's documentation regarding exposing scopes to clients. Each scope will contain claims that can be used to map to roles within workspaces in SystemLink. See [**Mapping OpenID Connect Claims to SystemLink Workspaces and Roles**](#mapping-openid-connect-claims-to-system ...
````
Might be worth mentioning that the `profile` and `email` scopes provide the claims SystemLink uses for the user's first/last name and email.

prestwick replied:
Good idea. Done!

chris468 commented on handbook/openid-connect/openid-connect.md:
````diff
@@ -0,0 +1,133 @@
+[protocol]://[systemlink-dns]/login/openidc-redirect
+
+```
+
````
Were you going to add something about the id_token encryption/signing settings that we've tested?

prestwick replied:
Yes, this was added since you provided this comment.

prestwick commented on handbook/chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,42 @@
+- *Are there common command line operations that should be used/understood?*
+- *How do I get started with installing or setting up any required tools?*
+
+### Example formatting CLI, configuration and code. 
```
Fixed

prestwick commented on handbook/chapter-template/chapter-template.md:
````diff
@@ -0,0 +1,42 @@
+systemlinkcli --apply
+```
+
+Below the snippet should be additional details regarding the result of the operations and the arguments, key/pairs, functions, etc that can/must be included to complete the operation. This should be several sentences long and leave the user with a complete understanding. Fenced code blocks should always be used when introducing the command/configuration/code, and the code block should be marked u ...
````
fixed!

prestwick commented on handbook/chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,42 @@
+Below the snippet should be additional details regarding the result of the operations and the arguments, key/pairs, functions, etc that can/must be included to complete the operation. This should be several sentences long and leave the user with a complete understanding. Fenced code blocks should always be used when introducing the command/configuration/code, and the code block should be marked u ...
+
+
+### In-line images
```
Fixed!

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,133 @@
+1. Log into the server running SystemLink and open **NI Web Server Configuration**
+2. Go to the **Authentication** tab and enable **Use OpenID Connect (advanced)**
+
+![Enable OpenID Connect in NI Web Server](odic-webserver.png)
```
fixed!

prestwick commented on handbook/chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,42 @@
+
+### In-line images
+
+Screen shots are encouraged, but be aware they may go stale as the product changes. Regardless they are both useful to make stress a point within our documentation, as well as make the overall content easier to read as it breaks up large walls of text. 
```
fixed!

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,133 @@
+![Enable OpenID Connect in NI Web Server](odic-webserver.png)
+*Enable OpenID Connect in NI Web Server*
+
+## OpenID Configuration Files in SystemLink server
```
Server should be capitalized here, but not every instance. I review the doc again to make sure this was correct in other uses.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,133 @@
+
+> For example, a OpenID provider with the DNS `openid-provider.com` would yield files named `openid-provider.com/conf` , `openid-provider.com.client`, and `openid-provider.com.provider`. 
+
+These files do no exist for new SystemLink installations. Each file must be added to the `C:\Program Files\National Instruments\Shared\Web Server\conf\openidc` directory. The NI Web Server must be restarted for these changes to take effect.
```
fixed!

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,133 @@
+# Authenticating with OpenID Connect
```
I like the suggestion. I went ahead and changed the title.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,133 @@
+
+Once SystemLink and the provider have been configured such that users can authenticate and login, mappings between OpenID claims and workspaces must be made in order to provide the user access to systems and data managed by SystemLink. This process is the same as the mapping workflow for LDAP and Active Directory attributes, groups, and users. Please see [Assigning Users to Roles in a Workspace]( ...
+
+The OpenID Connect provider determines which scopes and claims are made available to clients. The available claims can be viewed at the `userinfo_endpoint` hosted by the provider. Use `https://[provider-dns]/.well-known/openid-configuration` to determine the URL of the `userinfo_endpoint`. You will need to obtain a valid bearer token to authenticate and access this endpoint.  
```
Good call. I incorporated your example in place of mine.

prestwick replied:
I had added some of this info since you added this comment, but I was able to improve what I had here based on your feedback.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,133 @@
+
+## OpenID Configuration Files in SystemLink server
+
+There are three files that must be created to connect your SystemLink server to an OpenID Connect provider: `[provider-dns].conf`, `[provider-dns].client`, and `[provider-dns].provider`. The `[provider-dns]` portion of each filename must be the URL encoded fully qualified domain name. 
```
Good call. I incorporated your example in place of mine.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,133 @@
+
+Once SystemLink and the provider have been configured such that users can authenticate and login, mappings between OpenID claims and workspaces must be made in order to provide the user access to systems and data managed by SystemLink. This process is the same as the mapping workflow for LDAP and Active Directory attributes, groups, and users. Please see [Assigning Users to Roles in a Workspace]( ...
+
+The OpenID Connect provider determines which scopes and claims are made available to clients. The available claims can be viewed at the `userinfo_endpoint` hosted by the provider. Use `https://[provider-dns]/.well-known/openid-configuration` to determine the URL of the `userinfo_endpoint`. You will need to obtain a valid bearer token to authenticate and access this endpoint.  
```
I like how you worded this. I basically used your comment here in place of what I had originally.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,133 @@
+- A OpenID Connect Provider server such as [PingFederate](https://www.pingidentity.com/en/software/pingfederate.html), [Azure ADFS](https://docs.microsoft.com/en-us/windows-server/identity/ad-fs/deployment/how-to-connect-fed-azure-adfs), [Okta](https://www.okta.com/openid-connect/), or another [certified provider](https://openid.net/certification/) that has been fully setup and configured for Ope ...
+    - If you have not yet setup your provider please consult the vendor documentation for setup and configuration. 
+
+## Enabling OpenID Connect in SystemLink
```
This is a good idea. I will find a way to weave this in

prestwick replied:
Done!

prestwick commented on handbook/openid-connect/openid-connect.md:
````diff
@@ -0,0 +1,133 @@
+
+```
+The `client_id` and `client_secret` can be obtained directly from the provider. Depending on the provider the `client_id` may be user defined. After setting these properties the NI Web Server must be restarted for changes to take effect. 
+
````
Included a discussion of this point after the example `50_mod_auth_openidc-defines.conf` file.

spanglerco commented:
Was the favicon changed from the SystemLink logo to the generic NI intentionally?

spanglerco commented on mkdocs.yml:
```diff
@@ -39,3 +39,8 @@ theme:
 
 extra_css:
   - stylesheets/extra.css
+
+
+nav:
+  - 'README.md'
+  - 'openid-connect/openid-connect.md'
```
Instead of adding nav, you would edit [handbook/.pages](https://github.com/ni/systemlink-operations-handbook/blob/master/handbook/.pages).

spanglerco commented on operation-handbook.code-workspace:
```diff
@@ -0,0 +1,12 @@
+{
```
Is it normal practice to check in a file like this into source control?

prestwick replied:
I've added it to the gitignore and will remove it. That was an oversight.

prestwick replied:
removed!

chris468 commented on handbook/chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,42 @@
+# Chapter Template title
```
The build is warning about how they're in the build dir but not in the "nav" configuration

prestwick replied:
Moved!

chris468 commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,223 @@
+1. Log into the server running SystemLink and open **NI Web Server Configuration**
+2. Go to the **Authentication** tab and enable both **Log in as users controlled by the web server** and **Use OpenID Connect (advanced)**
+3. Add the configuration files to SystemLink to connect to your OpenID Connect provider and confirm that you can log in as an openid connect user
+> **Note:** details on this process is found in [**OpenID Configuration Files in SystemLink Server**](#openid-configuration-files-in-systemlink-server)
```
These look a little funny, the Note text is lighter than the other content, except the link is still dark. Maybe a we can tweak the css a bit.

![image](https://user-images.githubusercontent.com/62968881/98415108-3be92e80-2042-11eb-8be1-22940135e28d.png)

prestwick replied:
This is addressed by using admonitions here rather than block quotes.

prestwick commented:
@spanglerco I changed the favicon initially due to the issue described at https://stackoverflow.com/questions/59233531/safari-favicon-incorrectly-rendering-with-white-background. 

I've reverted this change after testing with more browsers. The NI icon doesn't get the weird white outline because it has enough contrast with the dark mode tabs in Safari.

prestwick commented on handbook/README.md:
```diff
@@ -11,8 +11,3 @@ Source and contributions for the handbook and example files can be found in the
-
-## Table of Contents
-
-- [Placeholder](placeholder) - a placeholder chapter that will be filled in
-  later.
```
Since we have an TOC that gets built automatically and is shown in the same page this felt redundant. I opted for removing this here since it will get increasingly brittle as we add more chapters.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,223 @@
+# Single Sign-on with OpenID Connect
+
+You may configure SystemLink to use [OpenID Connect](https://openid.net) for user authorization. This enables SystemLink to leverage corporate single sign-on (SSO) and the additional security benefits it provides such streamlined login and limiting the proliferation of user credentials. This also enables SystemLink to utilize a common identity for users across multiple applications. OpenID Connec ...
```
Fixed!

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,223 @@
+- A server running SystemLink 2020R4 or greater. Please see [Installing and Configuring SystemLink Server and Clients](https://www.ni.com/documentation/en/systemlink/latest/setup/configuring-systemlink-server-clients/) for the basics of setting up a SystemLink server. 
+- A DNS name for the SystemLink server. 
+- SystemLink login with the **Server Administrator** role. 
+- Administrator desktop access to the SystemLink server
```
Fixed!

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,223 @@
+1. Log into the server running SystemLink and open **NI Web Server Configuration**
+2. Go to the **Authentication** tab and enable both **Log in as users controlled by the web server** and **Use OpenID Connect (advanced)**
+3. Add the configuration files to SystemLink to connect to your OpenID Connect provider and confirm that you can log in as an openid connect user
+> **Note:** details on this process is found in [**OpenID Configuration Files in SystemLink Server**](#openid-configuration-files-in-systemlink-server)
```
Fixed! I will also be looking into a markdown linter once I fix all these grammatical errors.

prestwick commented on handbook/openid-connect/openid-connect.md:
````diff
@@ -0,0 +1,223 @@
+```
+*Example curl request to return user info. The bearer token has been truncated for readability.*
+
+Alternatively you can view claims returned by a particular user by modifying the httpd configuration on your SystemLink server. This is accomplished by adding `AUTH_OIDC_ENABLE_CLAIM_INFO` to `C:\Program Files\National Instruments\Shared\Web Server\conf\defines.d\50_mod_auth_openidc-defines.conf` and restarting NI Web Server. User claims can then be viewed at `protocol]://[systemlink-dns]/login/o ...
````
Fixed! I also updated this paragraph to be a order list workflow to improve readability.

prestwick commented on handbook/openid-connect/openid-connect.md:
````diff
@@ -0,0 +1,223 @@
+# Whether OIDC is enabled.
+Define AUTH_OIDC_ENABLED
+```
+*An example `50_mod_auth_openidc-defines.conf` modified to expose user claim. You must be logged via OpenID Connect to receive data this endpoint.*
````
Good suggestion. I'm using this psuedo caption approach for images, but I'm using admonitions as you suggest as well as an alternative to block quotes elsewhere in the document.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,223 @@
+
+If the provider is https with a certificate signed by a CA not included in the NI-CURL CA bundle (`C:\Program Files\National Instruments\Shared\nicurl\ca-bundle.crt`), then the `AUTH_OIDC_PROVIDER_DIR` define in `50_mod_auth_openidc-defines.conf` must to be updated to point to a CA bundle containing the provider's CA. The path can be absolute, or relative to `C:\Program Files\National Instruments ...
+
+### Mapping Claims to SystemLink Roles. 
```
fixed!

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,223 @@
+    - If you have not yet setup your provider please consult the vendor documentation for setup and configuration. 
+
+## Enabling OpenID Connect in SystemLink
+1. Log into the server running SystemLink and open **NI Web Server Configuration**
```
Addressed inconsistencies throughout the document.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,223 @@
+
+## Enabling OpenID Connect in SystemLink
+1. Log into the server running SystemLink and open **NI Web Server Configuration**
+2. Go to the **Authentication** tab and enable both **Log in as users controlled by the web server** and **Use OpenID Connect (advanced)**
```
Good suggestion. I've updated this workflow.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,223 @@
+8. To enable OpenID Connect as the only login option, go back to **NI Web Server Configuration**> **Authentication** and disable **Log in as users controlled by the web server**. 
+
+
+![Enable OpenID Connect in NI Web Server](odic-webserver.png){: style="height:500px;width:500px"}
```
fixed! great catch.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,223 @@
+
+> For example, a OpenID provider with the DNS `example.com:9999` would yield files named `oexample.com%3a9999.conf` , `example.com%3a9999.client`, and `example.com%3a9999.provider`. 
+
+These files do not exist for new SystemLink installations. Each file must be added to the `C:\Program Files\National Instruments\Shared\Web Server\conf\openidc` directory. The NI Web Server must be restarted for these changes changes to any of these files to take effect. 
```
fixed!

prestwick commented on handbook/openid-connect/openid-connect.md:
````diff
@@ -0,0 +1,223 @@
+```
+
+### Setting Login Redirect URI
+The client configuration for your provider requires a redirect URL that is used during the login flow. This must include the protocol (`https://` or `http:// `) and the DNS name of the SystemLink server. If the SystemLink server's DNS changes, this setting must be updated with the provider. 
````
I incorporated this feedback into this paragraph.

spanglerco commented on chapter-template/chapter-template.md:
````diff
@@ -19,24 +18,33 @@ This section should describe the tools or application that must be in place in o
 
 ```bash
 systemlinkcli --apply
 ```
````
Do you think it would make sense to include the code snippet inside the admonition as well? That way it groups the description with the snippet nicely. As in:

!!! note ""
    The following shows how to apply using the `systemlinkcli`.

    ```bash
    systemlinkcli --apply
    ```

prestwick replied:
I'm picking up what you're putting down ;-). I've updated how I'm using admonitions to incorporate your suggestion.

spanglerco commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -1,48 +1,71 @@
-    - If you have not yet setup your provider please consult the vendor documentation for setup and configuration. 
+- An OpenID Connect Provider server such as [PingFederate](https://www.pingidentity.com/en/software/pingfederate.html), [Azure ADFS](https://docs.microsoft.com/en-us/windows-server/identity/ad-fs/deployment/how-to-connect-fed-azure-adfs), [Okta](https://www.okta.com/openid-connect/), or another [certified provider](https://openid.net/certification/) that has been fully setup and configured for Op ...
+
+  - If you have not yet setup your provider please consult the vendor documentation for setup and configuration.
```
It looks like mkdocs wants four spaces in order to make it a nested list. You can configure markdownlint to do this in the .markdownlint.yaml:

```yaml
MD007:
  indent: 4
```

prestwick replied:
I went ahead and incorporated this rule.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,223 @@
+You may configure SystemLink to use [OpenID Connect](https://openid.net) for user authorization. This enables SystemLink to leverage corporate single sign-on (SSO) and the additional security benefits it provides such streamlined login and limiting the proliferation of user credentials. This also enables SystemLink to utilize a common identity for users across multiple applications. OpenID Connec ...
+
+
+
```
@spanglerco  I addressed linting errors locally. With you sign-off I'd like to defer getting a linter setup for pr builds. I'll do that in a subsequent PR. I don't want that to hold up this PR at this point.

spanglerco commented on .markdownlint.yaml:
```diff
@@ -0,0 +1,2 @@
+MD033: false
```
Consider whitelisting which elements to allow:

```suggestion
MD033:
  allowed_elements:
    - figure
    - figcaption
    - img
```

prestwick replied:
Glad to know we can do this! Added.

spanglerco commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -1,48 +1,71 @@
-SystemLink can be configured to support multiple OpenID Providers simultaneously by creating a `[provider-dns].conf`, `[provider-dns].client`, and `[provider-dns].provider` file for each provider.  It is required that the user ID used within SystemLink be unique across providers. This id takes the form `[sub_claim]@issuer` where this issuer is the DNS name of the OpenID Connect provider. You can  ...
+There are three files that must be created to connect your SystemLink server to an OpenID Connect provider: `[provider-dns].conf`, `[provider-dns].client`, and `[provider-dns].provider`. The `[provider-dns]` portion of each filename must be the URL encoded fully qualified domain name.
+
+!!! note ""
+    For example, a OpenID provider with the DNS `example.com:9999` would yield files named `example.com%3a9999.conf` , `example.com%3a9999.client`, and `example.com%3a9999.provider`.
```
Should we consider using the example admonition instead of the text "For example"? As in:

```suggestion
!!! example
    An OpenID provider with the DNS `example.com:9999` would yield files named `example.com%3a9999.conf` , `example.com%3a9999.client`, and `example.com%3a9999.provider`.
```

![image](https://user-images.githubusercontent.com/7519484/98724711-c17a2080-2359-11eb-9337-a0ad283a305e.png)

That way it's easy to scan through and tell what's an example vs. a note. Unless I guess we're intending this to be a clarification rather than an example?

prestwick replied:
Good suggestion. I went ahead and did this. I also updated the CSS for admonitions to support our theme as well.

spanglerco commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -1,48 +1,71 @@
+
+!!! note ""
+    For example, a OpenID provider with the DNS `example.com:9999` would yield files named `example.com%3a9999.conf` , `example.com%3a9999.client`, and `example.com%3a9999.provider`.
+]
```
Stray ] character here.
```suggestion

```

prestwick replied:
Fixed!

spanglerco commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -183,14 +243,23 @@ Define AUTH_OIDC_REDIRECT_URI /login/openidc-redirect
 
 # Whether OIDC is enabled.
 Define AUTH_OIDC_ENABLED
+AUTH_OIDC_ENABLE_CLAIM_INFO
```
This line isn't needed. It's the Define below that's correct.
```suggestion

```

prestwick replied:
Fixed!

jjaco2 commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+
+### Refreshing user claims
+
+If a claim for a user changes and that claim is used to drive a role mapping for a workspace, the user must logout and log back in before the new claim is received and the role assigned to the user changed.
```
`logout`
'log out'

Also, I don't understand this part: `before the new claim is received and the role assigned to the user changed.`

Is this saying there's like a narrow window of time before someone's mapping gets all messed up? If so, is there more detail we can give here or elsewhere to help prevent that?

prestwick replied:
Reworded this, but worth another look from you.

jjaco2 commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+
+## Troubleshooting Failed Authentication
+
+Due to the amount of configuration required by both the provider and SystemLink it can be challenging to enable the entire OpenID Connect flow. The following sources can be used to troubleshoot a failed connection.
```
I'd cut the first sentence for tone and minimalism and just jump to 'Use the following sources to troubleshoot...'

prestwick replied:
Changed.

jjaco2 reviewed (commented):
Clarity and concision edits, plus a few workflow questions. I'd like to review this again closer to release if possible, just to check for sanity with the other docs.

prestwick commented on CONTRIBUTING.md:
```diff
@@ -56,6 +56,11 @@ To access the preview, open a browser and navigate to <http://127.0.0.1:8000/>.
+## Authoring Handbook Chapters
+
+- Review **chapter-template.md** and use as a starting point and reference to the structure and style of the handbook.
+- To add a chapter you must update the `arrange` section in `handbook/.pages` in this repository to reference the new chapter directory.
```
fixed

prestwick commented on chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,50 @@
+# Chapter Template title
+
+Within the title section should be a brief description and overview of the goal to be achieved with the workflows and documentation described in the chapter. Please note, this template is to provide guidance on the format for chapters, but this format can and should be changed depending on the topic.
```
fixed!

prestwick commented on chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,50 @@
+
+## Assumptions and Prerequisites
+
+This section should describe the tools or application that must be in place in order for the workflow to be successfully executed.
```
changed

prestwick commented on chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,50 @@
+# Chapter Template title
+
+Within the title section should be a brief description and overview of the goal to be achieved with the workflows and documentation described in the chapter. Please note, this template is to provide guidance on the format for chapters, but this format can and should be changed depending on the topic.
```
changed

prestwick commented on chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,50 @@
+Within the title section should be a brief description and overview of the goal to be achieved with the workflows and documentation described in the chapter. Please note, this template is to provide guidance on the format for chapters, but this format can and should be changed depending on the topic.
+
+- *What goals will be achieved?*
+- *What questions will be answered?*
```
changed

prestwick commented on chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,50 @@
+- *What state or configuration must SystemLink be in?*
+- *Are there any licensing requirements that must be met?*
+- *Are there third party applications that must be in place?*
+- *Are there tools that must be installed to complete the workflows*
```
I'm going to leave this more verbose. Otherwise I worry that answering the specifics of these will be brushed over.

prestwick commented on chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,50 @@
+
+## Section N
+
+The remainder of the chapter should be broken down in the sections and subsections as needed. These should describe the workflow to be completed as well as the theory or underlying details that provide context to why the workflow is setup as it.
```
changed

prestwick commented on chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,50 @@
+
+The remainder of the chapter should be broken down in the sections and subsections as needed. These should describe the workflow to be completed as well as the theory or underlying details that provide context to why the workflow is setup as it.
+
+- *Are there common command line operations that should be used/understood?*
```
changed

prestwick commented on chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,50 @@
+
+### Example formatting CLI, configuration and code
+
+If needed lead into a code snip with a one line admonition. Keep in mind these do not render in Github style markdown, but will be rendered correctly in the built MkDocs site.
```
Changed

> Regarding the second 2 sentences -- are we saying to use markdown so that it shows up in the mkdocs site correctly, even if it looks weird in github?

Yes, we are going to have to accept some of that.

prestwick commented on chapter-template/chapter-template.md:
````diff
@@ -0,0 +1,50 @@
+systemlinkcli --apply
+```
+
+Below the snippet should be additional details regarding the result of the operations and the arguments, key/pairs, functions, etc. that can/must be included to complete the operation. This should be several sentences long and leave the user with a complete understanding. Fenced code blocks should always be used when introducing the command/configuration/code, and the code block should be marked  ...
````
Incorporated. 

My expectation is the doc developer will know what language to use.

prestwick commented on chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,50 @@
+
+Screen shots are encouraged, but be aware they may go stale as the product changes. Regardless they are both useful to stress a point within our documentation, as well as make the overall content easier to read as it breaks up large walls of text.
+
+Using inline HTML is recommended by MkDocs.
```
I made this more clear that Inline HTML is needed to get captions.

prestwick commented on chapter-template/chapter-template.md:
```diff
@@ -0,0 +1,50 @@
+
+### Inline images
+
+Screen shots are encouraged, but be aware they may go stale as the product changes. Regardless they are both useful to stress a point within our documentation, as well as make the overall content easier to read as it breaks up large walls of text.
```
Changed!

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+# Single Sign-on with OpenID Connect
+
+You may configure SystemLink to use [OpenID Connect](https://openid.net) for user authorization. This enables SystemLink to leverage corporate single sign-on (SSO) and the additional security benefits it provides such as streamlined login and limiting the proliferation of user credentials. This also enables SystemLink to utilize a common identity for users across multiple applications. OpenID Con ...
```
Changed!

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+
+- A server running SystemLink 2020R4 or greater. Please see [Installing and Configuring SystemLink Server and Clients](https://www.ni.com/documentation/en/systemlink/latest/setup/configuring-systemlink-server-clients/) for the basics of setting up a SystemLink server.
+
+- A DNS name for the SystemLink server.
```
Fixed! I left the periods in my workflow instructions (order lists) because I believe they are full sentences. Let me know if their lack of subject make me incorrect. 
e.g 
> 1. Log into the server running SystemLink and go to `C:\Program Files\National Instruments\Shared\Web Server\conf\openidc`.
>2. Add the configuration files to SystemLink to connect to your OpenID Connect provider.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+
+- An OpenID Connect Provider server such as [PingFederate](https://www.pingidentity.com/en/software/pingfederate.html), [Azure ADFS](https://docs.microsoft.com/en-us/windows-server/identity/ad-fs/deployment/how-to-connect-fed-azure-adfs), [Okta](https://www.okta.com/openid-connect/), or another [certified provider](https://openid.net/certification/) that has been fully setup and configured for Op ...
+
+  - If you have not yet setup your provider please consult the vendor documentation for setup and configuration.
```
True. Removed.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+2. Add the configuration files to SystemLink to connect to your OpenID Connect provider.
+
+    !!! note ""
+        Details on this process is found in [**OpenID Configuration Files in SystemLink Server**](#openid-configuration-files-in-systemlink-server).
```
Changed

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+
+5. Click **Apply and restart**.
+
+6. Log into the SystemLink web application using OpenID connect. After logging in you should see the message **There are no available applications.** because there are no role mappings yet created for this user.
```
Removed

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+
+6. Log into the SystemLink web application using OpenID connect. After logging in you should see the message **There are no available applications.** because there are no role mappings yet created for this user.
+
+7. Logout and log back into the SystemLink web application as the NI Web Server admin user. This is the user created during NI Web Server guided setup.
```
I changed this step up some. It's not important that it's the user created during guided setup. It just needs to be a user with the **Server Administrator** role. 

6. Login into the SystemLink web application with a user assigned the [**Server Administrator**](https://www.ni.com/documentation/en/systemlink/latest/setup/predefined-roles/) role.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+9. Add an OpenID Claim mapping for the **Server Administrator** role.
+
+    !!! note ""
+        Details on mapping claims to roles is found in [**Mapping OpenID Connect Claims to SystemLink Workspaces and Roles**](#mapping-openid-connect-claims-to-systemlink-workspaces-and-roles).
```
changed

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+    !!! note ""
+        Details on mapping claims to roles is found in [**Mapping OpenID Connect Claims to SystemLink Workspaces and Roles**](#mapping-openid-connect-claims-to-systemlink-workspaces-and-roles).
+
+10. Log in as an OpenID connect user with a mapping for the **Server Administrator** role and confirm they have the correct privileges.
```
Yes. Changed.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+
+## OpenID Configuration Files in SystemLink Server
+
+There are three files that must be created to connect your SystemLink server to an OpenID Connect provider: `[provider-dns].conf`, `[provider-dns].client`, and `[provider-dns].provider`. The `[provider-dns]` portion of each filename must be the URL encoded fully qualified domain name.
```
Changed

Changed

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+
+### Viewing Claims Returned by a Provider
+
+The OpenID Connect provider determines which scopes and claims are made available to clients. The available claims can be viewed at the `userinfo_endpoint` hosted by the provider. Use `https://[provider-dns]/.well-known/openid-configuration` to determine the URL of the `userinfo_endpoint`. You will need to obtain a valid bearer token to authenticate and access this endpoint.  
```
Changed

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+!!! note ""
+    For example, a OpenID provider with the DNS `example.com:9999` would yield files named `example.com%3a9999.conf` , `example.com%3a9999.client`, and `example.com%3a9999.provider`.
+]
+These files do not exist for new SystemLink installations. Each file must be added to the `C:\Program Files\National Instruments\Shared\Web Server\conf\openidc` directory. The NI Web Server must be restarted for any changes to these file to take effect.
```
changed

changed

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+]
+These files do not exist for new SystemLink installations. Each file must be added to the `C:\Program Files\National Instruments\Shared\Web Server\conf\openidc` directory. The NI Web Server must be restarted for any changes to these file to take effect.
+
+SystemLink can be configured to support multiple OpenID Providers simultaneously by creating a `[provider-dns].conf`, `[provider-dns].client`, and `[provider-dns].provider` file for each provider.  It is required that the user ID used within SystemLink be unique across providers. This id takes the form `[sub_claim]@issuer` where this issuer is the DNS name of the OpenID Connect provider. You can  ...
```
Changed

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+
+In this example the `openid`, `email`, and `profile` scopes are requested. Additional scopes may be requested. Consult your provider's documentation regarding exposing scopes to clients. The `profile` and `email` scopes are required to populate first name, last name, and email fields in the SystemLink user preferences.
+
+Each scope will contain claims that can be used to map to roles within workspaces in SystemLink. See [**Mapping OpenID Connect Claims to SystemLink Workspaces and Roles**](#mapping-openid-connect-claims-to-systemlink-workspaces-and-roles) for details.
```
Changed

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+
+### OpenID Configuration and Discovery
+
+The `[provider-dns].provider` file includes the contents of the provider's OpenID Connect configuration. This file informs SystemLink of of the various endpoints exposed by the provider that are used during the login process.
```
Changed

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+
+### Setting Login Redirect URI
+
+The client configuration for your provider requires a redirect URL that is used during the login flow. This must by the fully qualified domain name, the protocol (`https://` or `http://`), and the port (if 80 or 443 are not used) of the SystemLink server. If the SystemLink server's DNS changes, this setting must be updated with the provider.
```
Changed. I used 'update this section with *your* provider.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+The client configuration for your provider requires a redirect URL that is used during the login flow. This must by the fully qualified domain name, the protocol (`https://` or `http://`), and the port (if 80 or 443 are not used) of the SystemLink server. If the SystemLink server's DNS changes, this setting must be updated with the provider.
+
+It also must include the port if using a non-default (80/443). We should probably note that it's highly recommended to use the same name for the redirect URI as is set for the preferred host name on the remote tab of the web server config utility.
+
```
This was copy/paste fail, but it reminded me to include some info about preferred host names. I've cleaned this up.

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+
+## Supported Signing and Encryption Algorithms
+
+ The following algorithms are supported by SystemLink for ID token signing, ID token key management encryption, and ID token content encryption.
```
Changed

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+
+## Mapping OpenID Connect Claims to SystemLink Workspaces and Roles
+
+Once SystemLink and the provider have been configured such that users can authenticate and login, mappings between OpenID claims and workspaces must be made in order to authorize the user access to systems and data managed by SystemLink. This process is the same as the mapping workflow for LDAP and Active Directory attributes, groups, and users. Please see [Assigning Users to Roles in a Workspace ...
```
Changed

prestwick commented on handbook/openid-connect/openid-connect.md:
````diff
@@ -0,0 +1,295 @@
+curl -s https://slsso-runtime.grl-us1.uat.k8s.com/idp/userinfo.openid -H 'Authorization: Bearer eyJhbGciOiJSUzI1NiI...zJVy2oLnrBmXTmpDRm499U4~'|python -m json.tool
+```
+
+Alternatively you can view claims returned by a particular user by modifying the httpd configuration on your SystemLink server.
````
Changed

prestwick commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,295 @@
+Claims are returned as a JSON object.  
+
+!!! note ""
+    Example response from `userinfo_endpoint`. Any of these claims may be used to map a user to a role in a workspace.
```
Changed

spanglerco commented on IAM/openid/README.md:
```diff
@@ -0,0 +1,5 @@
+# OpenID Connect Configuration File for SystemLink
```
What are these IAM files? Did you mean to submit them here? You can have files within the handbook chapter and link to them. They'll show up as file downloads.

prestwick replied:
So I want to begin establishing a pattern where the handbook provides examples in addition to pure documentation. I do want to get feedback on here to place them. I put them in their current location to get feedback. 

Would you recommend a structure more like 
```
handbook-
      |
       - openid-connect
                     |
                      - openid-connect.md
                      - example
                              |
                               - foo.povider
                               - foo.client
                               - foo.conf
```
I'll probably move all the images in the chapter to the img directory.

spanglerco replied:
Yeah, I had imagined something along those lines. Users can still browse the repo (or we could provide links to the repo), but those files will also be included in the mkdocs site and show up as file downloads when linked to.

prestwick replied:
Changed!

I also changed the .pages file at the root of the handbook to use `nav` rather than `arrange` because arrange is to be [deprecated](https://github.com/lukasgeiter/mkdocs-awesome-pages-plugin#arrange-pages). I also added a .pages file in the directory containing the example config files so that its README.md file doesn't show up in handbook TOC.

spanglerco commented on IAM/openid/slsso-runtime.com.provider:
```diff
@@ -0,0 +1,34 @@
+
+{
```
Extra blank line
```suggestion
{
```

prestwick replied:
fixed!

spanglerco commented on handbook/openid-connect/openid-connect.md:
````diff
@@ -53,32 +49,33 @@ You may configure SystemLink to use [OpenID Connect](https://openid.net) for use
+
+    ```json
+    {
+    "scope": "openid email profile",
+    "ni-attributes": {
+      "displayName": "Log in with PingFederate",
+      "iconUri": "/login/assets/pf.png"
+    }
````
These lost some indentation:
```suggestion
      "scope": "openid email profile",
      "ni-attributes": {
        "displayName": "Log in with PingFederate",
        "iconUri": "/login/assets/pf.png"
      }
```

prestwick replied:
fixed!

spanglerco commented on handbook/openid-connect/openid-connect.md:
````diff
@@ -89,15 +86,15 @@ The `ni-attributes` section determines the text and (optionally) an icon to be s
-```
+    ```json
+    {
+    "client_id": "slserver",
+    "client_secret": "4vFm89u07xaredactedredactedredactede2tjtsEGQhlLreLVjcyLA0"
````
These lost some indentation:
```suggestion
      "client_id": "slserver",
      "client_secret": "4vFm89u07xaredactedredactedredactede2tjtsEGQhlLreLVjcyLA0"
```

prestwick replied:
fixed!

spanglerco commented on handbook/openid-connect/openid-connect.md:
````diff
@@ -113,27 +110,31 @@ The `client_id` and `client_secret` can be obtained from the provider. Depending
-```url
-[protocol]://[systemlink-dns]/login/openidc-redirect
+!!! note ""
+    Use the following URL to configure your provider
````
Should this end in a period or colon?
```suggestion
    Use the following URL to configure your provider:
```

prestwick replied:
I added the colon, but I'm curious @jjaco2 's feedback on this.

spanglerco commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -197,20 +198,19 @@ It also must include the port if using a non-default (80/443). We should probabl
 ## Mapping OpenID Connect Claims to SystemLink Workspaces and Roles
 
-Once SystemLink and the provider have been configured such that users can authenticate and login, mappings between OpenID claims and workspaces must be made in order to authorize the user access to systems and data managed by SystemLink. This process is the same as the mapping workflow for LDAP and Active Directory attributes, groups, and users. Please see [Assigning Users to Roles in a Workspace ...
+Map OpenID claims to roles and workspaces so users can access systems and data managed by SystemLink. This process is the same as the mapping workflow for LDAP and Active Directory attributes. Refer to[Assigning Users to Roles in a Workspace](https://www.ni.com/documentation/en/systemlink/latest/setup/mapping-roles/) in the SystemLink manual. You can also use claims to create a mapping for the ** ...
```
Missing a space:
```suggestion
Map OpenID claims to roles and workspaces so users can access systems and data managed by SystemLink. This process is the same as the mapping workflow for LDAP and Active Directory attributes. Refer to [Assigning Users to Roles in a Workspace](https://www.ni.com/documentation/en/systemlink/latest/setup/mapping-roles/) in the SystemLink manual. You can also use claims to create a mapping for the **Server Administrator** role.
```

prestwick replied:
fixed!

spanglerco commented on handbook/openid-connect/openid-connect.md:
````diff
@@ -258,35 +256,55 @@ If the provider is https with a certificate signed by a CA not included in the N
-  <img src="../claim-mapping.png" width="500" />
-  <figcaption>Mapping the ni_employee claim to a workspace.</figcaption>
-</figure>
+    ```json
+      "userinfo": {
````
Is this missing an opening { or something to make it a complete JSON document?

prestwick replied:
Fixed!

chris468 commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -0,0 +1,313 @@
+
+    <figure>
+      <img src="../escaped-claim.png" width="500" />
+      <figcaption>Claims must be escaped with the / character.</figcaption>
```
```suggestion
      <figcaption>Claims must be escaped with the \ character.</figcaption>
```

prestwick replied:
fixed!

spanglerco commented on handbook/openid-connect/openid-connect.md:
```diff
@@ -43,14 +43,16 @@ You can configure SystemLink to use OpenID Connect to authorize users. This enab
 
 There are three files that you must create to connect your SystemLink server to an OpenID Connect provider: `[provider-dns].conf`, `[provider-dns].client`, and `[provider-dns].provider`. The `[provider-dns]` portion of each filename must be the URL-encoded fully qualified domain name.
 
+Refer to [example-slclient-config](https://github.com/ni/systemlink-operations-handbook/tree/master/handbook/openid-connect/example-slclient-config) for examples of each of these files.
```
Do you think it would be beneficial to link directly to the example files (within the handbook) instead of users needing to switch to GitHub? As in, you could do something like:

```md
Refer to [slsso-runtime.com.client](/openid-connect/example-slclient-config/slsso-runtime.com.client), [slsso-runtime.com.conf](/openid-connect/example-slclient-config/slsso-runtime.com.conf), and [slsso-runtime.com.provider](/openid-connect/example-slclient-config/slsso-runtime.com.provider) for examples of each of these files.
```

And each of the links will download the associated file.

spanglerco replied:
Otherwise, I think it defeats the purpose of having the files within the handbook directory and should instead be in an examples directory or something outside of the handbook directory.

prestwick replied:
This has been moved outside of the handbook because its best just to send the user directly to GitHub at this point.
